Added Environment variables propagation capability with all/filtered/no propagation. made it backward compatible and opt-in per processor.

kikokikok opened the pull request.

kikokikok commented:
fixes #607

kikokikok commented:
First implementation to validate the concept. 
Should I gate it with a new LOVELY_ENV_PROPAGATION_ENABLED and LOVELY_ENV_PROPAGATION_FILTER environment variables ?

Joibel commented:
I don't think what you've done here will work as you're expecting.

`cmd.Env` is not currently set by the existing code, so it would pick up environment variables from the processes environment. Those specified in the application object do NOT get transferred to cmp plugins, so the ones you've written in your original issue won't be present - or they would just work out of the box without your code change. The ones you put in the application get transferred, but they get prefixed with `ARGOCD_ENV_` which you're not stripping anywhere.

What I'd have probably implemented is a new **parameter** (which you weirdly can specify through environment variables) to transfer a bunch of `KEY=value` environment variables in. I haven't thought this through fully.

kikokikok commented:
Oh I am not done at all ! I am going through a refactoring to have the Generate method take multiple optional GenerateOption instances, same with the Execute part. 
I'm added two Lovely features, one bool to enable env propagation and another with the prefix based filtering of the env variables to propagate. Making it available to all processors and extensible through the GenerateOption and ExecuteOption struct.

kikokikok commented:
BTW just understood your comment, I finished the refactoring I wanted to introduce so that you guys have an extensibility mechanism for later using GenerateOptions and ExecuteOptions 
- Will now use it to implement the missing piece. Prefix mapper from ARGO_ENV_ to injected ENV if I understood you correctly.

kikokikok commented:
@Joibel I think it should be fine now. I have drastically simplified my approach after getting to know a bit better your code base. 
Introduced the Env variables mapping as a feature with LOVELY_ENV_PROPAGATION 
And added a test for helmfile to test proper mapping from ARGOCD_ENV_ prefixed env to expected one.

Joibel commented on test/preprocessing_bydir_fallback/env.txt:
```diff
@@ -1,4 +1,4 @@
-LOVELY_PREPROCESSORS: sed -i 's/minimal/foobar/g' *.yaml
```
Can you explain these sed test case changes?

kikokikok commented:
Hello @Joibel,

Anything you'd like me to change in the PR, let me know. I tried to minimize the impact as much as possible.
Thanks !

Joibel commented:
> Hello @Joibel,
> 
> Anything you'd like me to change in the PR, let me know. I tried to minimize the impact as much as possible. Thanks !

Nothing at the moment, I haven't had a chance to look over the PR yet. It's the week before kubecon, which means I'm quite busy, but I will take a look when I get a chance.

Joibel reviewed (approved):
LGTM, thank you for doing this

Joibel commented:
I am releasing 1.2.0 with this in.
